pkg/icingaredis/own_heartbeat.go

```diff
@@ -0,0 +1,92 @@
+package icingaredis
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/go-redis/redis/v8"
+	"github.com/icinga/icingadb/pkg/com"
+	"github.com/icinga/icingadb/pkg/logging"
+	"github.com/icinga/icingadb/pkg/utils"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"strconv"
+	"time"
+)
+
+const statSecs = 15 * 60
+
+// Stats of sync by kind for self-monitoring.
+var Stats = struct {
+	Config  *com.Stat `json:"config"`
+	State   *com.Stat `json:"state"`
+	History *com.Stat `json:"history"`
+	Overdue *com.Stat `json:"overdue"`
+}{
+	com.NewStat(statSecs),
+	com.NewStat(statSecs),
+	com.NewStat(statSecs),
+	com.NewStat(statSecs),
+}
+
+// haer abstracts HA to prevent package import cycles.
+type haer interface {
+	Responsible() bool
+}
+
+// OwnHeartbeat periodically writes heartbeats to a Redis stream.
+type OwnHeartbeat struct {
+	ctx    context.Context
+	client *Client
+	ha     haer
+	logger *logging.Logger
+}
+
+func NewOwnHeartbeat(ctx context.Context, client *Client, ha haer, logger *logging.Logger) *OwnHeartbeat {
+	oh := &OwnHeartbeat{ctx, client, ha, logger}
+
+	go oh.run()
+
+	return oh
+}
+
+func (oh *OwnHeartbeat) run() {
+	periodically := time.NewTicker(3 * time.Second)
+	defer periodically.Stop()
+
+	boolToIntStr := map[bool]string{false: "0", true: "1"}
+	wasResponsible := oh.ha.Responsible()
+	responsibleSince := time.Now()
+
+	for {
+		isResponsible := oh.ha.Responsible()
+		if isResponsible != wasResponsible {
+			wasResponsible = isResponsible
+			responsibleSince = time.Now()
+		}
+
+		statsJson, err := json.Marshal(&Stats)
+		if err != nil {
+			oh.logger.Errorw("Can't update own heartbeat", zap.Error(errors.WithStack(err)))
```

Comment:
The message itself sounds more like it would refer to a Redis error. This is really some error condition that should never happen, but the message should probably mention JSON in some way.